Class cast exceptions on error processing

## Situation

Uploading an empty sechub configuration (only Api version is set to "1.0" ) leads to an server error 

### Data
```
2023-04-18 15:31:54.854 ERROR 28038 --- [0.1-8443-exec-6] o.a.c.c.C.[Tomcat].[localhost]           : Exception Processing ErrorPage[errorCode=0, location=/error]

org.springframework.web.util.NestedServletException: Request processing failed; nested exception is java.lang.ClassCastException: class org.springframework.validation.ObjectError cannot be cast to class org.springframework.validation.FieldError (org.springframework.validation.ObjectError and org.springframework.validation.FieldError are in unnamed module of loader 'app')
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1014)
	at org.springframework.web.servlet.FrameworkServlet.doPost(FrameworkServlet.java:909)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:696)
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:883)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:779)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:227)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
	at org.springframework.boot.web.servlet.filter.ErrorPageSecurityFilter.doFilter(ErrorPageSecurityFilter.java:80)
	at org.springframework.boot.web.servlet.filter.ErrorPageSecurityFilter.doFilter(ErrorPageSecurityFilter.java:70)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:337)
	at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.invoke(FilterSecurityInterceptor.java:106)
	at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.doFilter(FilterSecurityInterceptor.java:81)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:346)
	at org.springframework.security.web.access.ExceptionTranslationFilter.doFilter(ExceptionTranslationFilter.java:122)
	at org.springframework.security.web.access.ExceptionTranslationFilter.doFilter(ExceptionTranslationFilter.java:116)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:346)
	at org.springframework.security.web.session.SessionManagementFilter.doFilter(SessionManagementFilter.java:87)
	at org.springframework.security.web.session.SessionManagementFilter.doFilter(SessionManagementFilter.java:81)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:346)
	at org.springframework.security.web.authentication.AnonymousAuthenticationFilter.doFilter(AnonymousAuthenticationFilter.java:109)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:346)
	at org.springframework.security.web.servletapi.SecurityContextHolderAwareRequestFilter.doFilter(SecurityContextHolderAwareRequestFilter.java:149)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:346)
	at org.springframework.security.web.savedrequest.RequestCacheAwareFilter.doFilter(RequestCacheAwareFilter.java:63)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:346)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:102)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:346)
	at org.springframework.security.web.authentication.logout.LogoutFilter.doFilter(LogoutFilter.java:103)
	at org.springframework.security.web.authentication.logout.LogoutFilter.doFilter(LogoutFilter.java:89)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:346)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:102)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:346)
	at org.springframework.security.web.context.SecurityContextPersistenceFilter.doFilter(SecurityContextPersistenceFilter.java:112)
	at org.springframework.security.web.context.SecurityContextPersistenceFilter.doFilter(SecurityContextPersistenceFilter.java:82)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:346)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:102)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:346)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:102)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:346)
	at org.springframework.security.web.FilterChainProxy.doFilterInternal(FilterChainProxy.java:221)
	at org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:186)
	at org.springframework.web.filter.DelegatingFilterProxy.invokeDelegate(DelegatingFilterProxy.java:354)
	at org.springframework.web.filter.DelegatingFilterProxy.doFilter(DelegatingFilterProxy.java:267)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
	at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:100)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:102)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:102)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
	at org.apache.catalina.core.ApplicationDispatcher.invoke(ApplicationDispatcher.java:711)
	at org.apache.catalina.core.ApplicationDispatcher.processRequest(ApplicationDispatcher.java:461)
	at org.apache.catalina.core.ApplicationDispatcher.doForward(ApplicationDispatcher.java:385)
	at org.apache.catalina.core.ApplicationDispatcher.forward(ApplicationDispatcher.java:313)
	at org.apache.catalina.core.StandardHostValve.custom(StandardHostValve.java:388)
	at org.apache.catalina.core.StandardHostValve.status(StandardHostValve.java:244)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:171)
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:92)
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:78)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:360)
	at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:399)
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:65)
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:891)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1784)
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)
	at org.apache.tomcat.util.threads.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1191)
	at org.apache.tomcat.util.threads.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:659)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.ClassCastException: class org.springframework.validation.ObjectError cannot be cast to class org.springframework.validation.FieldError (org.springframework.validation.ObjectError and org.springframework.validation.FieldError are in unnamed module of loader 'app')
	at com.mercedesbenz.sechub.server.core.ServerError.<init>(ServerError.java:38)
	at com.mercedesbenz.sechub.server.core.ServerErrorController.error(ServerErrorController.java:42)
	at jdk.internal.reflect.GeneratedMethodAccessor120.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.springframework.web.method.support.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:205)
	at org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest(InvocableHandlerMethod.java:150)
	at org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:117)
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandlerMethod(RequestMappingHandlerAdapter.java:895)
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:808)
	at org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:87)
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1071)
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:964)
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1006)
	... 75 common frames omitted
```

--- a/sechub-commons-model/src/main/java/com/mercedesbenz/sechub/commons/model/SecHubConfigurationModelValidationError.java
+++ b/sechub-commons-model/src/main/java/com/mercedesbenz/sechub/commons/model/SecHubConfigurationModelValidationError.java
@@ -43,9 +43,9 @@ public enum SecHubConfigurationModelValidationError {
 
     METADATA_LABEL_VALUE_TOO_LONG("Meta data label value length is too long."),
 
-    METADATA_TOO_MANY_LABELS("Too many meta data labels defined"),
+    METADATA_TOO_MANY_LABELS("Too many meta data labels defined!"),
 
-    METADATA_LABEL_KEY_CONTAINS_ILLEGAL_CHARACTERS("Meta data label key contains illegal characters"),
+    METADATA_LABEL_KEY_CONTAINS_ILLEGAL_CHARACTERS("Meta data label key contains illegal characters."),
 
     ;
 

--- a/sechub-commons-model/src/test/java/com/mercedesbenz/sechub/commons/model/SecHubConfigurationModelValidationErrorTest.java
+++ b/sechub-commons-model/src/test/java/com/mercedesbenz/sechub/commons/model/SecHubConfigurationModelValidationErrorTest.java
@@ -1,0 +1,32 @@
+package com.mercedesbenz.sechub.commons.model;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.EnumSource;
+
+class SecHubConfigurationModelValidationErrorTest {
+
+    @ParameterizedTest
+    @EnumSource(SecHubConfigurationModelValidationError.class)
+    void message_does_end_with_dot_or_exclamation_mark(SecHubConfigurationModelValidationError error) {
+        /* execute */
+        String defaultMessage = error.getDefaultMessage();
+
+        /* test */
+        String[] expectedEndings = new String[] { ".", "!" };
+        boolean endsWithExpected = false;
+        for (String expectedEnding : expectedEndings) {
+            endsWithExpected = defaultMessage.endsWith(expectedEnding);
+            if (endsWithExpected) {
+                break;
+            }
+
+        }
+
+        if (!endsWithExpected) {
+            fail("Message does not end with expected endings:" + expectedEndings + ":\n" + defaultMessage);
+        }
+    }
+
+}

--- a/sechub-integrationtest/src/main/java/com/mercedesbenz/sechub/integrationtest/api/AsUser.java
+++ b/sechub-integrationtest/src/main/java/com/mercedesbenz/sechub/integrationtest/api/AsUser.java
@@ -504,13 +504,18 @@ public class AsUser {
     }
 
     public UUID createJobAndReturnJobUUID(TestProject project, SecHubScanConfiguration config) {
+        String resultAsString = createJobAndReturnResultAsString(project, config);
+        return fetchJobUUID(resultAsString);
+    }
+
+    public String createJobAndReturnResultAsString(TestProject project, SecHubScanConfiguration config) {
         String projectId = project.getProjectId();
         config.setProjectId(projectId);
 
         String json = config.toJSON();
         String url = getUrlBuilder().buildAddJobUrl(projectId);
         String resultAsString = getRestHelper().postJson(url, json);
-        return fetchJobUUID(resultAsString);
+        return resultAsString;
     }
 
     private String createTargetURIForSechubConfiguration(IntegrationTestMockMode runMode, List<String> whites) {

--- a/sechub-integrationtest/src/test/java/com/mercedesbenz/sechub/integrationtest/scenario4/SecHubConfigurationFailuresScenario4IntTest.java
+++ b/sechub-integrationtest/src/test/java/com/mercedesbenz/sechub/integrationtest/scenario4/SecHubConfigurationFailuresScenario4IntTest.java
@@ -1,0 +1,108 @@
+// SPDX-License-Identifier: MIT
+package com.mercedesbenz.sechub.integrationtest.scenario4;
+
+import static com.mercedesbenz.sechub.integrationtest.api.TestAPI.*;
+import static com.mercedesbenz.sechub.integrationtest.scenario4.Scenario4.*;
+import static org.junit.Assert.*;
+
+import org.junit.Rule;
+import org.junit.Test;
+import org.springframework.http.HttpStatus;
+import org.springframework.web.client.HttpClientErrorException;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.json.JsonMapper;
+import com.fasterxml.jackson.databind.node.ArrayNode;
+import com.mercedesbenz.sechub.commons.core.FailableRunnable;
+import com.mercedesbenz.sechub.commons.model.SecHubCodeScanConfiguration;
+import com.mercedesbenz.sechub.commons.model.SecHubConfigurationMetaData;
+import com.mercedesbenz.sechub.commons.model.SecHubScanConfiguration;
+import com.mercedesbenz.sechub.integrationtest.api.IntegrationTestSetup;
+
+public class SecHubConfigurationFailuresScenario4IntTest {
+
+    @Rule
+    public IntegrationTestSetup setup = IntegrationTestSetup.forScenario(Scenario4.class);
+
+    @Test
+    public void create_project_missing_api_version_fails_with_HTTP_400_having_json_with_details() throws Exception {
+        /* prepare */
+        SecHubScanConfiguration config = new SecHubScanConfiguration();
+
+        /* execute + test */
+        JsonNode json = assertFailedWithJsonOutput(HttpStatus.BAD_REQUEST, () -> as(USER_1).createJobAndReturnResultAsString(PROJECT_1, config));
+        assertJsonContainsDetails(json, "Field 'apiVersion' with value 'null' was rejected. Api version is missing.");
+
+    }
+
+    @Test
+    public void create_project_api_version_set_but_nothing_else() throws Exception {
+        /* prepare */
+        SecHubScanConfiguration config = new SecHubScanConfiguration();
+        config.setApiVersion("1.0");
+
+        /* execute + test */
+        JsonNode json = assertFailedWithJsonOutput(HttpStatus.BAD_REQUEST, () -> as(USER_1).createJobAndReturnResultAsString(PROJECT_1, config));
+        assertJsonContainsDetails(json, "Configuration does not contain any scan option. Unable to start scan!");
+
+    }
+
+    @Test
+    public void create_project_api_version_set_but_labels_wrong() throws Exception {
+        /* prepare */
+        SecHubScanConfiguration config = new SecHubScanConfiguration();
+        config.setApiVersion("1.0");
+        SecHubCodeScanConfiguration codeScan = new SecHubCodeScanConfiguration();
+        config.setCodeScan(codeScan);
+
+        SecHubConfigurationMetaData metaData = new SecHubConfigurationMetaData();
+        config.setMetaData(metaData);
+        metaData.getLabels().put("E xample1.1", "value");
+
+        /* execute + test */
+        JsonNode json = assertFailedWithJsonOutput(HttpStatus.BAD_REQUEST, () -> as(USER_1).createJobAndReturnResultAsString(PROJECT_1, config));
+        assertJsonContainsDetails(json,
+                "Meta data label key contains illegal characters. Label key 'E xample1.1' may only contain 'a-z','0-9', '-', '_' or '.' characters");
+
+    }
+
+    private void assertJsonContainsDetails(JsonNode json, String... details) {
+        ArrayNode detailsNode = (ArrayNode) json.get("details");
+
+        int index = 0;
+
+        for (String detail : details) {
+
+            JsonNode detailNode = detailsNode.get(index);
+
+            String detailText = detailNode.asText();
+            if (!detailText.contains(detail)) {
+                // use now assertEquals to have text compare editors inside IDE on test
+                // runtime...
+                assertEquals(detail, detailText);
+            }
+
+            index++;
+        }
+
+    }
+
+    private JsonNode assertFailedWithJsonOutput(HttpStatus expectedStatus, FailableRunnable<Exception> failable) throws Exception {
+        try {
+            failable.runOrFail();
+            fail("Shall be not reachable - configuration was wrong. Must have 400 failure");
+
+        } catch (HttpClientErrorException clientException) {
+            assertEquals(expectedStatus, clientException.getStatusCode());
+            String response = clientException.getResponseBodyAsString();
+            try {
+                JsonMapper mapper = JsonMapper.builder().build();
+                return mapper.reader().readTree(response);
+            } catch (Exception e) {
+                fail("The response is not valid json:\n" + response);
+            }
+        }
+        return null; // unnecessary, but otherwise compile error...
+    }
+
+}

--- a/sechub-server-core/src/main/java/com/mercedesbenz/sechub/server/core/ServerError.java
+++ b/sechub-server-core/src/main/java/com/mercedesbenz/sechub/server/core/ServerError.java
@@ -5,7 +5,10 @@ import java.util.ArrayList;
 import java.util.List;
 import java.util.Map;
 
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 import org.springframework.validation.FieldError;
+import org.springframework.validation.ObjectError;
 
 import com.fasterxml.jackson.annotation.JsonAutoDetect;
 import com.fasterxml.jackson.annotation.JsonAutoDetect.Visibility;
@@ -19,6 +22,8 @@ import com.mercedesbenz.sechub.commons.model.JSONable;
 @JsonAutoDetect(fieldVisibility = Visibility.ANY, getterVisibility = Visibility.NONE, setterVisibility = Visibility.NONE)
 @JsonIgnoreProperties(ignoreUnknown = true)
 public class ServerError implements JSONable<ServerError> {
+    private static final Logger LOG = LoggerFactory.getLogger(ServerError.class);
+
     Integer status;
     String error;
 
@@ -32,18 +37,38 @@ public class ServerError implements JSONable<ServerError> {
         this.error = (String) errorAttributes.get("error");
         this.message = (String) errorAttributes.get("message");
 
-        @SuppressWarnings("unchecked")
-        List<FieldError> list = (List<FieldError>) errorAttributes.getOrDefault("errors", new ArrayList<>());
-        this.details = new ArrayList<>();
-        for (FieldError fieldError : list) {
-            StringBuilder sb = new StringBuilder();
-            sb.append("Field '").append(fieldError.getField());
-            sb.append("' with value '" + fieldError.getRejectedValue());
-            sb.append("' was rejected. " + fieldError.getDefaultMessage());
-            details.add(sb.toString());
-        }
         this.timeStamp = errorAttributes.get("timestamp").toString();
         this.trace = (String) errorAttributes.get("trace");
+
+        this.details = new ArrayList<>();
+
+        Object errorObjects = errorAttributes.getOrDefault("errors", new ArrayList<>());
+
+        if (errorObjects instanceof List) {
+            List<?> list = (List<?>) errorObjects;
+
+            for (Object object : list) {
+                if (object instanceof FieldError) {
+                    FieldError fieldError = (FieldError) object;
+
+                    StringBuilder sb = new StringBuilder();
+                    sb.append("Field '").append(fieldError.getField());
+                    sb.append("' with value '" + fieldError.getRejectedValue());
+                    sb.append("' was rejected. " + fieldError.getDefaultMessage());
+                    details.add(sb.toString());
+                } else if (object instanceof ObjectError) {
+                    ObjectError objectError = (ObjectError) object;
+                    details.add(objectError.getDefaultMessage());
+                } else {
+                    LOG.warn("Untreated object inside errors found:{}", object);
+                }
+
+            }
+        } else {
+            if (errorObjects != null) {
+                LOG.error("Error objects is not a list or null but: {}", errorObjects);
+            }
+        }
     }
 
     @Override

--- a/sechub-server-core/src/main/java/com/mercedesbenz/sechub/server/core/ServerErrorController.java
+++ b/sechub-server-core/src/main/java/com/mercedesbenz/sechub/server/core/ServerErrorController.java
@@ -28,7 +28,7 @@ public class ServerErrorController implements ErrorController {
 
     @MustBeDocumented("When debug flag is set, rest call reponse error messages do also contain stacktraces.")
     @Value("${sechub.server.debug:false}")
-    private boolean debug;
+    boolean debug;
 
     @Autowired
     ErrorAttributes errorAttributes;
@@ -68,6 +68,7 @@ public class ServerErrorController implements ErrorController {
 
             if (httpStatusEnum.is4xxClientError()) {
                 options = options.including(ErrorAttributeOptions.Include.MESSAGE);
+                options = options.including(ErrorAttributeOptions.Include.BINDING_ERRORS);
             }
 
         } catch (RuntimeException e) {

--- a/sechub-server-core/src/test/java/com/mercedesbenz/sechub/server/core/ServerErrorControllerTest.java
+++ b/sechub-server-core/src/test/java/com/mercedesbenz/sechub/server/core/ServerErrorControllerTest.java
@@ -4,6 +4,11 @@ package com.mercedesbenz.sechub.server.core;
 import static org.junit.jupiter.api.Assertions.*;
 import static org.mockito.Mockito.*;
 
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.List;
+
 import javax.servlet.RequestDispatcher;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
@@ -12,6 +17,10 @@ import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
 import org.springframework.boot.web.servlet.error.DefaultErrorAttributes;
 import org.springframework.http.ResponseEntity;
+import org.springframework.validation.BindException;
+import org.springframework.validation.BindingResult;
+import org.springframework.validation.FieldError;
+import org.springframework.validation.ObjectError;
 
 class ServerErrorControllerTest {
 
@@ -33,7 +42,7 @@ class ServerErrorControllerTest {
     }
 
     @Test
-    void using_default_errorattributes_when_response_is_401_the_returns_json_body_contains_the_error_message() {
+    void when_response_is_401_the_json_body_contains_the_error_message() {
         /* prepare */
         prepareError(401, TEST_ERROR_MESSAGE);
 
@@ -47,7 +56,94 @@ class ServerErrorControllerTest {
     }
 
     @Test
-    void using_default_errorattributes_when_response_is_501_the_returns_json_body_contains_NOT_the_error_message() {
+    void when_response_is_401_the_json_body_contains_the_error_message_and_binding_details1() {
+        /* prepare */
+        FieldError objectError = new FieldError("errorObjectName", "fieldName", "defaultMessageToResolveThis");
+
+        simulateBindingExceptionOnSprintBootServerSide("bindingResultObjectName", objectError);
+
+        prepareError(401, "something");
+
+        /* execute */
+        ResponseEntity<ServerError> serverErrorResponse = controllerToTest.error(request, response);
+
+        /* test */
+        ServerError serverError = serverErrorResponse.getBody();
+        assertEquals("Validation failed for object='bindingResultObjectName'. Error count: 1", serverError.message);
+
+        assertEquals(1, serverError.details.size());
+        assertEquals("Field 'fieldName' with value 'null' was rejected. defaultMessageToResolveThis", serverError.details.iterator().next());
+
+    }
+
+    @Test
+    void when_response_is_401_the_json_body_contains_the_error_message_and_binding_details2_check_no_stacktrace() {
+        /* prepare */
+        FieldError objectError1 = new FieldError("errorObjectName1", "fieldName1", "defaultMessageToResolveThis1");
+        FieldError objectError2 = new FieldError("errorObjectName2", "fieldName2", "defaultMessageToResolveThis2");
+
+        simulateBindingExceptionOnSprintBootServerSide("bindingResultObjectName", objectError1, objectError2);
+
+        prepareError(401, "something");
+
+        /* execute */
+        ResponseEntity<ServerError> serverErrorResponse = controllerToTest.error(request, response);
+
+        /* test */
+        ServerError serverError = serverErrorResponse.getBody();
+        assertEquals("Validation failed for object='bindingResultObjectName'. Error count: 2", serverError.message);
+
+        assertEquals(2, serverError.details.size());
+        Iterator<String> iterator = serverError.details.iterator();
+        assertEquals("Field 'fieldName1' with value 'null' was rejected. defaultMessageToResolveThis1", iterator.next());
+        assertEquals("Field 'fieldName2' with value 'null' was rejected. defaultMessageToResolveThis2", iterator.next());
+
+        assertNull(serverError.trace);
+    }
+
+    @Test
+    void when_response_is_401_the_json_body_contains_the_error_message_and_binding_details2_check_with_stacktrace_when_debug() {
+        /* prepare */
+        FieldError objectError1 = new FieldError("errorObjectName1", "fieldName1", "defaultMessageToResolveThis1");
+        FieldError objectError2 = new FieldError("errorObjectName2", "fieldName2", "defaultMessageToResolveThis2");
+
+        simulateBindingExceptionOnSprintBootServerSide("bindingResultObjectName", objectError1, objectError2);
+
+        prepareError(401, "something");
+        controllerToTest.debug = true;
+
+        /* execute */
+        ResponseEntity<ServerError> serverErrorResponse = controllerToTest.error(request, response);
+
+        /* test */
+        ServerError serverError = serverErrorResponse.getBody();
+        assertEquals("Validation failed for object='bindingResultObjectName'. Error count: 2", serverError.message);
+
+        assertEquals(2, serverError.details.size());
+        Iterator<String> iterator = serverError.details.iterator();
+        assertEquals("Field 'fieldName1' with value 'null' was rejected. defaultMessageToResolveThis1", iterator.next());
+        assertEquals("Field 'fieldName2' with value 'null' was rejected. defaultMessageToResolveThis2", iterator.next());
+
+        assertNotNull(serverError.trace);
+    }
+
+    private void simulateBindingExceptionOnSprintBootServerSide(String objectName, FieldError... objectErrors) {
+        BindingResult bindingResult = mock(BindingResult.class);
+
+        List<ObjectError> allErrors = new ArrayList<>();
+
+        allErrors.addAll(Arrays.asList(objectErrors));
+
+        when(bindingResult.getObjectName()).thenReturn(objectName);
+        when(bindingResult.getAllErrors()).thenReturn(allErrors);
+        when(bindingResult.getErrorCount()).thenReturn(allErrors.size());
+
+        BindException bindException = new BindException(bindingResult);
+        when(request.getAttribute(RequestDispatcher.ERROR_EXCEPTION)).thenReturn(bindException);
+    }
+
+    @Test
+    void when_response_is_501_the_json_body_contains_NOT_the_error_message() {
         /* prepare */
         prepareError(501, TEST_ERROR_MESSAGE);
 

